### ISSUE TITLE
Transmission-daemon 2.92 slow performance with large torrent

### ISSUE THREAD

**Issue by Jorman** [transmission/transmission#728]
Hi, I've a nas with debian 9.5 and I installed transmission via apt

I don't have to much torrent but are very big
![2](https://user-images.githubusercontent.com/5203988/46333001-e9491700-c61e-11e8-8b9c-001daea28b92.PNG)

The problem is that transmission became slow, 1 time of 5 go timeout to the remote gui and the speed was not full speed, even the upload are very low, I can upload up to 1.5mb but don't go...
Even the transmission-remote -l via ssh sometime give timeout

All the torrent finish to one usb3.0 external hd mounted ntfs.

![1](https://user-images.githubusercontent.com/5203988/46332905-85bee980-c61e-11e8-97ff-e0976c462030.PNG)

What I can do?
There's any way to compile it in a better and optimized way to my Atom processor? Maybe using all 4 core

**Comment by AndreyPavlenko** [comment]
I also observe performance degradation on large files. On the begging the download speed is close to my network bandwidth, but it gets lower and lower while downloading large files.
I don't think it's caused by high CPU usage, on my amlogic912-based device the transmission process consumes ~ 5% CPU.

**Comment by Jorman** [comment]
This happens only on very large torrent, never happen before even with 10 and more torrent in seeding. I know that some people use transmission with more than a hundred torrent, ok the have a better cpu, but like I said, if I download 1 torrent with a lot of sources no problem, if I download one very large torrent with a lot of sources I have a lot of problems of speed and performance
I tried a different client, qbittorrent, and no problem of speed or performance degradation with the large torrents

**Comment by OMICRON3069** [comment]
Well, I test it by add few of torrent which contains large file, and it actually causes low performance on my seed box.

I haven't try other client like qbittorrent because of lake of basic feature(e.g. tracker list modification) for remote user.

**Comment by chchia** [comment]
I also find that my transmission 2.94 become very slow when large size torrent in the list, the remote gui will get timeout frequently even if no much of bandwidth used.

**Comment by jtl999** [comment]
I've also noticed the same with >1TB torrents.

Transmission/2.94+

**Comment by jtl999** [comment]
I'm busy, but a suggestion to start with would be for someone knowledgeable to use something like [Valgrind](http://www.valgrind.org/) to profile Transmission and find out what's hogging CPU time when downloading a torrent with a large torrent.

**Comment by cfpp2p** [comment]
@jtl999 
Related to [TRAC-1753](https://trac.transmissionbt.com/ticket/1753)
See also https://github.com/transmission/transmission/issues/108 and the like.

**Comment by Jorman** [comment]
Can be a kind of hash problem?
I mean, more the torrent is big, more the parts are big, 4 or 8 mib or even more instead 256 or 512kib for smaller torrent
Can this cause the constant hash check by transmission?
So the time to check the hash for 8mib parts is bigger than the one to check the 512kib and the overall performance drop down and even the other torrent can not be downloaded at full speed

Can be possible?

**Comment by AndreyPavlenko** [comment]
I doubt it is related to the hash because at the beginning the speed is ok, but it degrades over the time.

**Comment by jtl999** [comment]
Hmm.

Testing on my own server now.

LXC container with 2 cores of a Xeon E3, 2GB RAM and writing to a RAID10 ZFS array that can do >200MB/s peak roughly.

Started downloading the whole of a certain 1.6TB torrent, transmission download speed very slow <1MB/s (My setup is normally capable of >100Mbps) and CPU usage is high.

![](https://i.imgur.com/OOKV6vi.png)

Select a small fraction of the same torrent so I'm only downloading about 10GB.

No high CPU usage:

Download speed good considering number of seeders/peers, etc.

![](https://i.imgur.com/0XtkVGA.png)

**Comment by jtl999** [comment]
Hmm.

Testing on my own server now.

LXC container with 2 cores of a Xeon E3, 2GB RAM and writing to a RAID10 ZFS array that can do >200MB/s peak roughly.

Started downloading the whole of a certain 1.6TB torrent, transmission download speed very slow <1MB/s (My setup is normally capable of >100Mbps) and CPU usage is high.

![](https://i.imgur.com/OOKV6vi.png)

Select a small fraction of the same torrent so I'm only downloading about 10GB.

CPU usage of transmission process is only about 20-40% but otherwise no problem.

Download speed good considering number of seeders/peers, etc.

![](https://i.imgur.com/64HghlD.png)

**Comment by zappepappe** [comment]
I'm having similar issues on a FreeNAS jail using Transmission 2.94.
Decided to try and see what Transmission was doing so ran some tests with DTrace (complete newbie with DTrace). Doesn't tell me much, but hopefully it will help someone else figure out what's going on. Each trace ran for about 10 sec.

No download with some light uploading:
```
Connections:	Down: 0		Up: 2-4
Bandwidth:	Down: 0		Up: ~100 KB/s
CPU Load:	~2%

# ./hotuser -p `pgrep -n transmission-daemon`
Sampling... Hit Ctrl-C to end.
^C
FUNCTION                                                COUNT   PCNT
transmission-daemon`0x445909                                1   0.7%
libcrypto.so.8`0x801193643                                  1   0.7%
libdht.so.0`0x801b01a32                                     1   0.7%
transmission-daemon`0x405814                                1   0.7%
transmission-daemon`0x405a74                                1   0.7%
transmission-daemon`0x44c97a                                1   0.7%
transmission-daemon`0x425c80                                1   0.7%
transmission-daemon`0x425cbe                                1   0.7%
transmission-daemon`0x44448c                                1   0.7%
libc.so.7`0x8023fafd1                                       1   0.7%
transmission-daemon`0x43229f                                1   0.7%
libthr.so.3`0x802709c64                                     1   0.7%
transmission-daemon`0x425ac2                                1   0.7%
transmission-daemon`0x40eae3                                1   0.7%
transmission-daemon`0x40b085                                1   0.7%
transmission-daemon`0x425af9                                1   0.7%
libcrypto.so.8`0x80118455c                                  1   0.7%
libc.so.7`0x80242f5e6                                       1   0.7%
libc.so.7`0x8023f419b                                       1   0.7%
libevent-2.1.so.6.0.2`0x8008b3ae2                           1   0.7%
transmission-daemon`0x425aea                                1   0.7%
transmission-daemon`0x425d71                                1   0.7%
transmission-daemon`0x435b82                                1   0.7%
transmission-daemon`0x40af9b                                1   0.7%
transmission-daemon`0x4245f4                                1   0.7%
libcrypto.so.8`BN_mod_mul_montgomery                        1   0.7%
libcrypto.so.8`0x8011936d9                                  1   0.7%
libc.so.7`vsnprintf                                         1   0.7%
transmission-daemon`0x425ae4                                1   0.7%
transmission-daemon`0x445a07                                1   0.7%
libcrypto.so.8`0x8011935e1                                  1   0.7%
libc.so.7`0x80242eab0                                       1   0.7%
libc.so.7`strlcpy                                           1   0.7%
libc.so.7`0x80243106c                                       1   0.7%
transmission-daemon`0x412558                                1   0.7%
libc.so.7`0x8024996b9                                       1   0.7%
libevent-2.1.so.6.0.2`evbuffer_add                          1   0.7%
transmission-daemon`0x445182                                1   0.7%
transmission-daemon`0x4249fb                                1   0.7%
libcrypto.so.8`0x80118446b                                  1   0.7%
transmission-daemon`0x44572e                                1   0.7%
transmission-daemon`0x425a69                                1   0.7%
libc.so.7`pthread_equal                                     1   0.7%
libevent-2.1.so.6.0.2`evbuffer_invoke_callbacks_            1   0.7%
libevent-2.1.so.6.0.2`0x8008b3bb8                           1   0.7%
libthr.so.3`_pthread_self                                   1   0.7%
transmission-daemon`0x44598a                                1   0.7%
transmission-daemon`0x43236b                                1   0.7%
transmission-daemon`0x40afc3                                1   0.7%
transmission-daemon`0x42464a                                1   0.7%
libevent-2.1.so.6.0.2`0x8008b3db4                           1   0.7%
transmission-daemon`0x428dd7                                1   0.7%
libcrypto.so.8`RC4                                          1   0.7%
transmission-daemon`0x424646                                1   0.7%
libcrypto.so.8`BN_div                                       1   0.7%
transmission-daemon`0x432292                                1   0.7%
transmission-daemon`0x44c970                                1   0.7%
transmission-daemon`0x40af87                                1   0.7%
libc.so.7`0x8023d9b5e                                       1   0.7%
libc.so.7`__sys_kevent                                      1   0.7%
libcrypto.so.8`bn_add_words                                 1   0.7%
transmission-daemon`0x424740                                1   0.7%
transmission-daemon`0x40ead1                                1   0.7%
libcrypto.so.8`bn_sqr_normal                                1   0.7%
transmission-daemon`0x432387                                1   0.7%
transmission-daemon`0x43b408                                1   0.7%
libc.so.7`memcpy                                            2   1.4%
libevent-2.1.so.6.0.2`evbuffer_add_vprintf                  2   1.4%
transmission-daemon`0x432366                                2   1.4%
transmission-daemon`0x428ce8                                2   1.4%
transmission-daemon`0x424864                                2   1.4%
transmission-daemon`0x428cb4                                2   1.4%
transmission-daemon`0x435b64                                2   1.4%
transmission-daemon`0x417969                                3   2.2%
transmission-daemon`0x428de0                                3   2.2%
libcrypto.so.8`bn_mul_add_words                             3   2.2%
transmission-daemon`0x43235d                                3   2.2%
libc.so.7`qsort                                             4   2.9%
libc.so.7`strcmp                                            4   2.9%
transmission-daemon`0x4458e3                                4   2.9%
transmission-daemon`0x432360                                5   3.6%
transmission-daemon`0x417944                                5   3.6%
transmission-daemon`0x432363                                7   5.1%
transmission-daemon`0x432355                                8   5.8%
libc.so.7`memcmp                                            9   6.5%

# ./procsystime -n transmission-daemon
Tracing... Hit Ctrl-C to end...
dtrace: 32 dynamic variable drops with non-empty dirty list
dtrace: 45 dynamic variable drops with non-empty dirty list
dtrace: 501 dynamic variable drops with non-empty dirty list
dtrace: 37 dynamic variable drops with non-empty dirty list
dtrace: 43 dynamic variable drops with non-empty dirty list
dtrace: 128 dynamic variable drops with non-empty dirty list
dtrace: 558 dynamic variable drops with non-empty dirty list
dtrace: 1107 dynamic variable drops with non-empty dirty list
dtrace: 135 dynamic variable drops with non-empty dirty list
^C
dtrace: 134 dynamic variable drops with non-empty dirty list

Elapsed Times for processes transmission-daemon,

         SYSCALL          TIME (ns)
            read               5538
            poll               8042
           write               9266
          openat              30004
            bind              30965
          socket              56921
           fcntl              79916
          getpid              90168
   posix_fadvise             103466
      setsockopt             110925
            stat             144764
        recvfrom             170134
         connect             175005
           fstat             189874
          accept             200935
          sendto             401212
           ioctl             462280
           close             501013
           readv             763139
   clock_gettime             971247
    gettimeofday            1461877
          writev            2378724
           pread           98598670
          select         9820680234
          kevent        18081774372

```
Download of a 17 GB torrent with some light uploading:
```
Connections:	Down: 8		Up: 2
Bandwidth:	Down: ~10 MB/s	Up: ~100 KB/s
CPU Load:	25-30%

# ./hotuser -p `pgrep -n transmission-daemon`
Sampling... Hit Ctrl-C to end.
^C
FUNCTION                                                COUNT   PCNT
libevent-2.1.so.6.0.2`event_add                             1   0.0%
libevent-2.1.so.6.0.2`0x8008bedc1                           1   0.0%
libc.so.7`0x802498100                                       1   0.0%
libevent-2.1.so.6.0.2`0x8008cd192                           1   0.0%
libcrypto.so.8`0x8011935e1                                  1   0.0%
libthr.so.3`0x802709d45                                     1   0.0%
libthr.so.3`0x802709c04                                     1   0.0%
transmission-daemon`0x444848                                1   0.0%
libthr.so.3`0x80270e3ba                                     1   0.0%
libthr.so.3`0x80270a1c5                                     1   0.0%
transmission-daemon`0x448c4b                                1   0.0%
transmission-daemon`0x440e82                                1   0.0%
libc.so.7`0x8023fac57                                       1   0.0%
transmission-daemon`0x43220b                                1   0.0%
transmission-daemon`0x432143                                1   0.0%
transmission-daemon`0x4322a6                                1   0.0%
libthr.so.3`0x80270a1a6                                     1   0.0%
transmission-daemon`0x445951                                1   0.0%
transmission-daemon`0x44c977                                1   0.0%
libc.so.7`0x8024981ef                                       1   0.0%
libevent-2.1.so.6.0.2`0x8008c5e60                           1   0.0%
transmission-daemon`0x43586d                                1   0.0%
transmission-daemon`0x432204                                1   0.0%
libthr.so.3`0x802704cc4                                     1   0.0%
libc.so.7`0x8023f9f94                                       1   0.0%
libc.so.7`0x8023feed1                                       1   0.0%
transmission-daemon`0x43cb04                                1   0.0%
libdht.so.0`0x801b01624                                     1   0.0%
transmission-daemon`0x417957                                1   0.0%
libc.so.7`0x8023fadb1                                       1   0.0%
libc.so.7`0x8023fac34                                       1   0.0%
libc.so.7`0x80240fa0d                                       1   0.0%
libc.so.7`0x8023facda                                       1   0.0%
libc.so.7`0x8023f9a44                                       1   0.0%
libevent-2.1.so.6.0.2`event_initialized                     1   0.0%
transmission-daemon`0x434821                                1   0.0%
libc.so.7`0x8023fac7e                                       1   0.0%
libevent-2.1.so.6.0.2`evmap_io_add_                         1   0.0%
libevent-2.1.so.6.0.2`0x8008cd041                           1   0.0%
transmission-daemon`0x43cf9b                                1   0.0%
libc.so.7`0x8023fef19                                       1   0.0%
transmission-daemon`0x44c510                                1   0.0%
transmission-daemon`0x44c7a3                                1   0.0%
libthr.so.3`0x802709261                                     1   0.0%
transmission-daemon`0x4490ae                                1   0.0%
transmission-daemon`0x43176a                                1   0.0%
libthr.so.3`0x802704cc2                                     1   0.0%
transmission-daemon`0x431c01                                1   0.0%
transmission-daemon`0x4247aa                                1   0.0%
libc.so.7`0x802498189                                       1   0.0%
transmission-daemon`0x43f6f7                                1   0.0%
libevent-2.1.so.6.0.2`evbuffer_decref_and_unlock_           1   0.0%
transmission-daemon`0x445175                                1   0.0%
transmission-daemon`0x437533                                1   0.0%
libevent-2.1.so.6.0.2`0x8008c6d7d                           1   0.0%
libc.so.7`0x8023f41a0                                       1   0.0%
libc.so.7`0x8023fefc7                                       1   0.0%
libc.so.7`0x8023f9a55                                       1   0.0%
libevent-2.1.so.6.0.2`0x8008c5d98                           1   0.0%
libc.so.7`0x8023fab58                                       1   0.0%
libc.so.7`strlcpy                                           1   0.0%
libevent-2.1.so.6.0.2`0x8008ccf2d                           1   0.0%
libthr.so.3`0x80270a161                                     1   0.0%
libcrypto.so.8`0x8011bcc1c                                  1   0.0%
libc.so.7`0x8023f23a1                                       1   0.0%
transmission-daemon`0x448ae8                                1   0.0%
libc.so.7`0x8023f242e                                       1   0.0%
transmission-daemon`0x432119                                1   0.0%
libthr.so.3`0x80270fafc                                     1   0.0%
libevent-2.1.so.6.0.2`0x8008b2848                           1   0.0%
libc.so.7`0x8023fac3d                                       1   0.0%
transmission-daemon`0x421083                                1   0.0%
libdht.so.0`0x801b01537                                     1   0.0%
libc.so.7`0x80240f6c5                                       1   0.0%
transmission-daemon`0x42108f                                1   0.0%
transmission-daemon`0x445036                                1   0.0%
transmission-daemon`0x4323fb                                1   0.0%
transmission-daemon`0x4321ac                                1   0.0%
transmission-daemon`0x4210f4                                1   0.0%
transmission-daemon`0x442489                                1   0.0%
transmission-daemon`0x444892                                1   0.0%
transmission-daemon`0x4441d4                                1   0.0%
libc.so.7`0x8024980f5                                       1   0.0%
libevent-2.1.so.6.0.2`0x8008c6d7a                           1   0.0%
libc.so.7`0x8023fa1f8                                       1   0.0%
libcrypto.so.8`0x801193617                                  1   0.0%
libc.so.7`0x8023f40f2                                       1   0.0%
transmission-daemon`0x43ccb7                                1   0.0%
libc.so.7`0x8023fa392                                       1   0.0%
libc.so.7`0x8023d9c99                                       1   0.0%
libcrypto.so.8`bn_mul_words                                 1   0.0%
transmission-daemon`0x444205                                1   0.0%
transmission-daemon`0x448c35                                1   0.0%
libthr.so.3`0x802709dcf                                     1   0.0%
transmission-daemon`0x417938                                1   0.0%
transmission-daemon`0x428cd8                                1   0.0%
transmission-daemon`0x40b07b                                1   0.0%
libevent-2.1.so.6.0.2`0x8008b182c                           1   0.0%
transmission-daemon`0x437cd9                                1   0.0%
transmission-daemon`0x448b09                                1   0.0%
libthr.so.3`0x802709cf7                                     1   0.0%
transmission-daemon`0x445465                                1   0.0%
transmission-daemon`0x40a80e                                1   0.0%
libc.so.7`0x8023f9f9d                                       1   0.0%
libc.so.7`0x8023facdf                                       1   0.0%
transmission-daemon`0x431954                                1   0.0%
libcrypto.so.8`0x801193545                                  1   0.0%
libthr.so.3`0x802709c01                                     1   0.0%
transmission-daemon`0x40a864                                1   0.0%
transmission-daemon`0x425d7e                                1   0.0%
transmission-daemon`0x44c95a                                1   0.0%
transmission-daemon`0x4459f0                                1   0.0%
transmission-daemon`0x43f5e8                                1   0.0%
libc.so.7`0x8024980fb                                       1   0.0%
libevent-2.1.so.6.0.2`0x8008ccf22                           1   0.0%
transmission-daemon`0x4320da                                1   0.0%
transmission-daemon`0x424f24                                1   0.0%
transmission-daemon`0x41d7a5                                1   0.0%
libevent-2.1.so.6.0.2`evbuffer_remove_buffer                1   0.0%
libc.so.7`localeconv_l                                      1   0.0%
transmission-daemon`0x43d0f3                                1   0.0%
libc.so.7`0x8023fac73                                       1   0.0%
libc.so.7`0x8023fac49                                       1   0.0%
transmission-daemon`0x43234c                                1   0.0%
transmission-daemon`0x4458d4                                1   0.0%
libcrypto.so.8`0x8011935ae                                  1   0.0%
transmission-daemon`0x434840                                1   0.0%
libthr.so.3`0x80270a12d                                     1   0.0%
libthr.so.3`0x802709d5f                                     1   0.0%
transmission-daemon`0x448d67                                1   0.0%
libcrypto.so.8`0x80119370d                                  1   0.0%
transmission-daemon`0x434838                                1   0.0%
transmission-daemon`0x440eb1                                1   0.0%
libevent-2.1.so.6.0.2`evbuffer_peek                         1   0.0%
libc.so.7`0x8023f420c                                       1   0.0%
transmission-daemon`0x44c9a2                                1   0.0%
libevent-2.1.so.6.0.2`0x8008b2814                           1   0.0%
libc.so.7`0x8023fa1a7                                       1   0.0%
transmission-daemon`0x40b03c                                1   0.0%
transmission-daemon`0x43cfce                                1   0.0%
libthr.so.3`0x802709c34                                     1   0.0%
transmission-daemon`0x444f8c                                1   0.0%
transmission-daemon`0x40a0c6                                1   0.0%
transmission-daemon`0x440bfd                                1   0.0%
libevent-2.1.so.6.0.2`evbuffer_add                          1   0.0%
transmission-daemon`0x40eaa1                                1   0.0%
transmission-daemon`0x432c26                                1   0.0%
libc.so.7`0x80242e959                                       1   0.0%
libc.so.7`0x8023fac67                                       1   0.0%
transmission-daemon`0x43747f                                1   0.0%
transmission-daemon`0x43d6de                                1   0.0%
transmission-daemon`0x4130c1                                1   0.0%
libthr.so.3`0x802709d03                                     1   0.0%
libevent-2.1.so.6.0.2`0x8008b3adb                           1   0.0%
libthr.so.3`0x80270fadc                                     1   0.0%
transmission-daemon`0x43f5cf                                1   0.0%
libevent-2.1.so.6.0.2`0x8008b60dd                           1   0.0%
libc.so.7`0x80249804c                                       1   0.0%
libthr.so.3`0x80270a0b3                                     1   0.0%
libevent-2.1.so.6.0.2`0x8008cd0f1                           1   0.0%
transmission-daemon`0x438dec                                1   0.0%
libthr.so.3`_pthread_self                                   1   0.0%
transmission-daemon`0x4290e7                                1   0.0%
libc.so.7`0x8023fa202                                       1   0.0%
libthr.so.3`0x802704882                                     1   0.0%
transmission-daemon`0x40a116                                1   0.0%
libc.so.7`0x8023fa8f0                                       1   0.0%
libc.so.7`0x802498179                                       1   0.0%
libevent-2.1.so.6.0.2`evbuffer_add_vprintf                  1   0.0%
libevent-2.1.so.6.0.2`evbuffer_reserve_space                1   0.0%
libc.so.7`_clock_gettime                                    1   0.0%
libc.so.7`0x8023f9d4c                                       1   0.0%
libevent-2.1.so.6.0.2`0x8008cd20d                           1   0.0%
libthr.so.3`0x80270e320                                     1   0.0%
libc.so.7`0x8023f20e1                                       1   0.0%
libc.so.7`0x8023fefe3                                       1   0.0%
transmission-daemon`0x448c72                                1   0.0%
libevent-2.1.so.6.0.2`0x8008bedd1                           1   0.0%
transmission-daemon`0x43236d                                1   0.0%
transmission-daemon`0x40b098                                1   0.0%
transmission-daemon`0x444871                                1   0.0%
libc.so.7`0x8023f9662                                       1   0.0%
libthr.so.3`0x802709c08                                     1   0.0%
transmission-daemon`0x40a88d                                1   0.0%
libc.so.7`0x80243089d                                       1   0.0%
libc.so.7`0x80240f631                                       1   0.0%
transmission-daemon`0x44c97d                                1   0.0%
transmission-daemon`0x434e3c                                1   0.0%
libthr.so.3`0x802709de1                                     1   0.0%
transmission-daemon`0x412331                                1   0.0%
libc.so.7`0x802497f91                                       1   0.0%
libthr.so.3`0x802704c71                                     1   0.0%
transmission-daemon`0x4247bf                                1   0.0%
transmission-daemon`0x444884                                1   0.0%
transmission-daemon`0x44c951                                1   0.0%
transmission-daemon`0x40a8ba                                1   0.0%
libevent-2.1.so.6.0.2`evbuffer_remove                       1   0.0%
libc.so.7`0x8023fb009                                       1   0.0%
libcrypto.so.8`BN_CTX_end                                   1   0.0%
libthr.so.3`0x80270fad1                                     1   0.0%
transmission-daemon`0x44c784                                1   0.0%
transmission-daemon`0x43f5ab                                1   0.0%
transmission-daemon`0x43b408                                1   0.0%
transmission-daemon`0x4406c6                                1   0.0%
transmission-daemon`0x4441c6                                1   0.0%
libevent-2.1.so.6.0.2`0x8008ccf19                           1   0.0%
transmission-daemon`0x437482                                1   0.0%
libc.so.7`0x8023fac7b                                       1   0.0%
libc.so.7`0x802499886                                       1   0.0%
transmission-daemon`0x4441c1                                1   0.0%
libthr.so.3`0x802709400                                     1   0.0%
transmission-daemon`0x425af2                                1   0.0%
transmission-daemon`0x4290a2                                1   0.0%
libc.so.7`0x8023fac83                                       1   0.0%
transmission-daemon`0x44c526                                1   0.0%
transmission-daemon`0x44c78d                                1   0.0%
transmission-daemon`0x440d9d                                2   0.1%
libevent-2.1.so.6.0.2`evbuffer_invoke_callbacks_            2   0.1%
transmission-daemon`0x43c9f1                                2   0.1%
transmission-daemon`0x43f481                                2   0.1%
transmission-daemon`0x434820                                2   0.1%
libevent-2.1.so.6.0.2`0x8008bed81                           2   0.1%
transmission-daemon`0x40a6d2                                2   0.1%
libevent-2.1.so.6.0.2`0x8008cce31                           2   0.1%
libc.so.7`0x8024981cc                                       2   0.1%
transmission-daemon`0x434842                                2   0.1%
libc.so.7`readv                                             2   0.1%
transmission-daemon`0x4444be                                2   0.1%
libcrypto.so.8`0x8010b2f81                                  2   0.1%
transmission-daemon`0x43199a                                2   0.1%
libc.so.7`0x8023faa54                                       2   0.1%
libevent-2.1.so.6.0.2`event_add_nolock_                     2   0.1%
libc.so.7`clock_gettime                                     2   0.1%
libthr.so.3`0x802709d5c                                     2   0.1%
libc.so.7`0x802498197                                       2   0.1%
transmission-daemon`0x44c798                                2   0.1%
libevent-2.1.so.6.0.2`evbuffer_read_setup_vecs_             2   0.1%
transmission-daemon`0x444841                                2   0.1%
transmission-daemon`0x4320d1                                2   0.1%
libc.so.7`0x8023f219c                                       2   0.1%
libevent-2.1.so.6.0.2`0x8008c6d9f                           2   0.1%
transmission-daemon`0x414751                                2   0.1%
transmission-daemon`0x448c3c                                2   0.1%
libc.so.7`qsort                                             2   0.1%
transmission-daemon`0x432d10                                3   0.1%
libthr.so.3`pthread_mutex_unlock                            3   0.1%
transmission-daemon`0x432360                                3   0.1%
libevent-2.1.so.6.0.2`evbuffer_commit_space                 3   0.1%
transmission-daemon`0x432363                                3   0.1%
libc.so.7`__sys_kevent                                      3   0.1%
transmission-daemon`0x43214f                                3   0.1%
libc.so.7`0x8023faa5d                                       3   0.1%
transmission-daemon`0x434830                                3   0.1%
libc.so.7`0x80249818c                                       3   0.1%
libc.so.7`__error                                           3   0.1%
transmission-daemon`0x434851                                3   0.1%
transmission-daemon`0x4206e1                                3   0.1%
transmission-daemon`0x4444c1                                3   0.1%
transmission-daemon`0x43f5df                                3   0.1%
libc.so.7`strcmp                                            3   0.1%
transmission-daemon`0x4458e3                                3   0.1%
libevent-2.1.so.6.0.2`evbuffer_add_buffer                   3   0.1%
transmission-daemon`0x432191                                3   0.1%
transmission-daemon`0x412463                                3   0.1%
transmission-daemon`0x417969                                4   0.2%
transmission-daemon`0x43235d                                4   0.2%
libc.so.7`bsearch                                           4   0.2%
libevent-2.1.so.6.0.2`evbuffer_write_atmost                 4   0.2%
libthr.so.3`pthread_equal                                   4   0.2%
transmission-daemon`0x41234b                                4   0.2%
transmission-daemon`0x43483b                                4   0.2%
libevent-2.1.so.6.0.2`evbuffer_read                         5   0.2%
libevent-2.1.so.6.0.2`evbuffer_get_length                   5   0.2%
transmission-daemon`0x432355                                5   0.2%
libevent-2.1.so.6.0.2`event_changelist_remove_all_          5   0.2%
libc.so.7`ioctl                                             5   0.2%
transmission-daemon`0x4448af                                5   0.2%
libc.so.7`0x8023f26a9                                       6   0.3%
transmission-daemon`0x432d1d                                6   0.3%
libevent-2.1.so.6.0.2`event_base_loop                       6   0.3%
transmission-daemon`0x43f5c7                                6   0.3%
transmission-daemon`0x418443                                6   0.3%
libc.so.7`memcmp                                            7   0.3%
libc.so.7`_readv                                            7   0.3%
libevent-2.1.so.6.0.2`evbuffer_ptr_set                      7   0.3%
libc.so.7`0x8023f4139                                       7   0.3%
libevent-2.1.so.6.0.2`event_callback_activate_nolock_        8   0.4%
libcrypto.so.8`EVP_EncryptUpdate                            8   0.4%
libcrypto.so.8`EVP_CipherUpdate                             9   0.4%
libc.so.7`0x8023fef1e                                       9   0.4%
transmission-daemon`0x43482a                               10   0.5%
libevent-2.1.so.6.0.2`evmap_io_active_                     13   0.6%
libevent-2.1.so.6.0.2`evbuffer_expand_fast_                15   0.7%
libevent-2.1.so.6.0.2`event_active_nolock_                 15   0.7%
libcrypto.so.8`EVP_DecryptUpdate                           16   0.7%
libevent-2.1.so.6.0.2`evbuffer_copyout_from                18   0.8%
libevent-2.1.so.6.0.2`0x8008b6162                          18   0.8%
libc.so.7`memmove                                          37   1.7%
libc.so.7`__malloc                                         38   1.8%
libc.so.7`memcpy                                           49   2.3%
libcrypto.so.8`sha1_block_data_order                      238  11.1%
libcrypto.so.8`RC4                                        275  12.8%
transmission-daemon`0x432d23                              295  13.7%
transmission-daemon`0x432d15                              309  14.4%
transmission-daemon`0x432d20                              320  14.9%

# ./procsystime -n transmission-daemon
Tracing... Hit Ctrl-C to end...
dtrace: 30320 dynamic variable drops with non-empty dirty list
dtrace: 31020 dynamic variable drops with non-empty dirty list
dtrace: 31927 dynamic variable drops with non-empty dirty list
dtrace: 29352 dynamic variable drops with non-empty dirty list
dtrace: 32462 dynamic variable drops with non-empty dirty list
dtrace: 31847 dynamic variable drops with non-empty dirty list
dtrace: 30776 dynamic variable drops with non-empty dirty list
dtrace: 30570 dynamic variable drops with non-empty dirty list
dtrace: 30701 dynamic variable drops with non-empty dirty list
dtrace: 31193 dynamic variable drops with non-empty dirty list
^C
dtrace: 32187 dynamic variable drops with non-empty dirty list

Elapsed Times for processes transmission-daemon,

         SYSCALL          TIME (ns)
          getuid                663
        quotactl                739
        pathconf               5636
            mmap               5700
            bind              12559
            poll              12857
          statfs              13265
           fstat              16098
        recvfrom              20228
          getpid              23106
   posix_fadvise              34280
          socket              44525
           fcntl              50391
       ftruncate              55118
      setsockopt              68383
         connect              97499
          accept             113186
          rename             152130
           close             211680
          openat             453107
          writev             821659
            stat            1073016
          sendto            1080062
         madvise            2068877
           fsync           11820563
           ioctl           15650913
           pread           17362059
   clock_gettime           22075564
    gettimeofday           23017151
           readv           35538432
          pwrite           59975825
          kevent         3106206333
          select         4342193469
```
Finally, downloading a 490 GB torrent and some light uploading:
```
Connections:	Down: 10	Up: 2-4
Bandwidth:	Down: ~3.5 MB/s	Up: ~100 KB/s
CPU Load:	~100%

# ./hotuser -p `pgrep -n transmission-daemon`
Sampling... Hit Ctrl-C to end.
^C
FUNCTION                                                COUNT   PCNT
libthr.so.3`0x8027049ce                                     1   0.0%
transmission-daemon`0x4441cb                                1   0.0%
libc.so.7`0x8023f2193                                       1   0.0%
transmission-daemon`0x43e4b9                                1   0.0%
libc.so.7`0x8023fa17d                                       1   0.0%
transmission-daemon`0x4490a6                                1   0.0%
libcrypto.so.8`0x80119b9de                                  1   0.0%
libc.so.7`0x8023fa18f                                       1   0.0%
transmission-daemon`0x431dee                                1   0.0%
libevent-2.1.so.6.0.2`evbuffer_read                         1   0.0%
transmission-daemon`0x434836                                1   0.0%
libthr.so.3`0x802709c8e                                     1   0.0%
libcrypto.so.8`EVP_DecodeBlock                              1   0.0%
transmission-daemon`0x442b10                                1   0.0%
libc.so.7`_readv                                            1   0.0%
libevent-2.1.so.6.0.2`evbuffer_drain                        1   0.0%
libevent-2.1.so.6.0.2`0x8008b6162                           1   0.0%
libcrypto.so.8`EVP_CipherUpdate                             1   0.0%
transmission-daemon`0x43b408                                1   0.0%
libc.so.7`0x8023fa58f                                       1   0.0%
transmission-daemon`0x41190b                                1   0.0%
transmission-daemon`0x417944                                1   0.0%
transmission-daemon`0x432cda                                1   0.0%
libssl.so.8`ssl3_read_bytes                                 1   0.0%
transmission-daemon`0x4176f5                                1   0.0%
libcrypto.so.8`EVP_DecodeUpdate                             1   0.0%
libevent-2.1.so.6.0.2`evbuffer_add_buffer                   1   0.0%
libthr.so.3`0x802709c2c                                     1   0.0%
libssl.so.8`ssl_cipher_ptr_id_cmp                           1   0.0%
libc.so.7`0x8023f9c82                                       1   0.0%
transmission-daemon`0x43f5e8                                1   0.0%
libc.so.7`0x8023f40fa                                       1   0.0%
libc.so.7`memchr                                            1   0.0%
libc.so.7`memcmp                                            1   0.0%
libc.so.7`0x8023fb155                                       1   0.0%
libc.so.7`0x8023f9cb9                                       1   0.0%
libc.so.7`0x8023faf4a                                       1   0.0%
libthr.so.3`0x802709400                                     1   0.0%
libc.so.7`0x8023fefc7                                       1   0.0%
transmission-daemon`0x414773                                1   0.0%
libc.so.7`0x8023fa76d                                       1   0.0%
libc.so.7`time                                              1   0.0%
transmission-daemon`0x445933                                1   0.0%
transmission-daemon`0x434840                                1   0.0%
libthr.so.3`0x802709721                                     1   0.0%
transmission-daemon`0x434830                                1   0.0%
libc.so.7`0x8023fef1e                                       1   0.0%
libc.so.7`0x8023f966d                                       1   0.0%
libthr.so.3`__pthread_cleanup_pop_imp                       1   0.0%
libc.so.7`0x8023f270d                                       1   0.0%
libthr.so.3`0x802709997                                     1   0.0%
libcrypto.so.8`0x8010b2f81                                  1   0.0%
libthr.so.3`0x8027049b4                                     1   0.0%
libc.so.7`0x802497ff9                                       1   0.0%
transmission-daemon`0x421099                                1   0.0%
libc.so.7`0x8023fad9d                                       1   0.0%
libc.so.7`0x8023f9a44                                       1   0.0%
transmission-daemon`0x44c7a3                                1   0.0%
libcrypto.so.8`0x8011b5f61                                  1   0.0%
transmission-daemon`0x43f5d3                                1   0.0%
transmission-daemon`0x43cf91                                1   0.0%
transmission-daemon`0x435861                                1   0.0%
libcrypto.so.8`ecp_nistz256_point_add                       1   0.0%
transmission-daemon`0x434827                                1   0.0%
libcrypto.so.8`0x8011b6443                                  1   0.0%
libthr.so.3`0x80270940d                                     1   0.0%
libevent-2.1.so.6.0.2`evbuffer_invoke_callbacks_            1   0.0%
transmission-daemon`0x432338                                1   0.0%
transmission-daemon`0x434e4a                                1   0.0%
transmission-daemon`0x44c984                                2   0.0%
libevent-2.1.so.6.0.2`evbuffer_get_length                   2   0.0%
libc.so.7`ioctl                                             2   0.0%
libevent-2.1.so.6.0.2`evbuffer_read_setup_vecs_             2   0.0%
transmission-daemon`0x4441c6                                2   0.0%
transmission-daemon`0x434820                                2   0.0%
libc.so.7`0x8023f4139                                       2   0.0%
transmission-daemon`0x432355                                2   0.0%
transmission-daemon`0x432c26                                2   0.0%
transmission-daemon`0x432363                                2   0.0%
libcrypto.so.8`EVP_DecryptUpdate                            2   0.0%
libevent-2.1.so.6.0.2`evbuffer_copyout_from                 3   0.0%
libc.so.7`memset                                            4   0.0%
transmission-daemon`0x432360                                4   0.0%
transmission-daemon`0x43483b                                4   0.0%
libc.so.7`memcpy                                           16   0.2%
transmission-daemon`0x43482a                               16   0.2%
libc.so.7`memmove                                          23   0.2%
transmission-daemon`0x432d10                               32   0.3%
libcrypto.so.8`RC4                                         64   0.7%
libcrypto.so.8`sha1_block_data_order                       67   0.7%
transmission-daemon`0x432d1d                               75   0.8%
transmission-daemon`0x432d20                             3043  31.3%
transmission-daemon`0x432d23                             3088  31.7%
transmission-daemon`0x432d15                             3200  32.9%

# ./procsystime -n transmission-daemon
Tracing... Hit Ctrl-C to end...
dtrace: 9 dynamic variable drops with non-empty dirty list
dtrace: 99 dynamic variable drops with non-empty dirty list
dtrace: 708 dynamic variable drops with non-empty dirty list
dtrace: 15 dynamic variable drops with non-empty dirty list
dtrace: 6 dynamic variable drops with non-empty dirty list
dtrace: 703 dynamic variable drops with non-empty dirty list
dtrace: 66 dynamic variable drops with non-empty dirty list
dtrace: 27 dynamic variable drops with non-empty dirty list
dtrace: 6 dynamic variable drops with non-empty dirty list
^C
dtrace: 135 dynamic variable drops with non-empty dirty list

Elapsed Times for processes transmission-daemon,

         SYSCALL          TIME (ns)
       issetugid               1490
      getsockopt               3058
     getsockname               3632
     getpeername               4371
            mmap               7844
          getpid              10346
           fcntl              10649
            bind              11022
      setsockopt              14139
        recvfrom              24171
   posix_fadvise              26164
            open              28768
           write              32566
          accept              35440
         thr_new              37980
          socket              42224
           fstat              49954
           close              84459
         connect              88654
          openat             138893
          sendto             157124
          writev             263667
            read             288327
         madvise             295492
   clock_gettime             395316
            stat             424132
            poll             513592
           ioctl           14403870
    gettimeofday           23761161
           readv           49653503
          pwrite          191898444
           pread          287734936
          kevent        10005647552
          select        10703422648
```

**Comment by hussion** [comment]
the same problem with transmission 2.94

**Comment by topstop3416** [comment]
I also encountered the same problem

**Comment by jcisio** [comment]
> Started downloading the whole of a certain 1.6TB torrent, transmission download speed very slow <1MB/s (My setup is normally capable of >100Mbps) and CPU usage is high.
> Select a small fraction of the same torrent so I'm only downloading about 10GB.
> CPU usage of transmission process is only about 20-40% but otherwise no problem.

I have been using this workaround for huge torrent and it works well. Downloading a 3 TB torrent at less than 1 MB/s and the UI is unresponsive. Selecting only 300 GB increases the speed to 3 MB and selecting only 100 GB can see an average download speed of 5-7 MB/s. When these 100 GB is almost done, selecting 100 GB more and the speed is always good. That way I can download huge torrent with good speed.

**Comment by UltimateByte** [comment]
I might be observing the same behavior on big files.

Analyzing machine usage with Netdata seems to indicate that downloading slows down or even stops when transmission-daemon is writing to the disk.

It's not writing as it downloads: it downloads a chunk, then writes it, slowing down the download.
To me, it seems like an issue with the way buffering is handled.
Not sure if this has always been the case, but I've noticed that since a few years as well. Might be related to changes in Linux kernel, or might be an issue since day one, and we're starting to notice as download speed gets faster, downloads gets bigger, but disk speed didn't improve much.

You can observe that behavior already on a medium sized 12GB download, but I confirm it seems worse on bigger torrents:
![netdata dl](https://user-images.githubusercontent.com/8805941/85619981-658d5480-b663-11ea-855d-48814c37edc1.jpg)
(pink is io wait, so system waiting on the disk to keep doing what it's doing)

I'm also unsure if it makes a difference to download one big file or multiple smaller files, this should be tested as well.

Hope this helps. I invite developers to check on what's going on with Netdata to hopefully help understanding what's really going on and if there is anything to improve in the Transmission server. And I'll try to remain available if there is anything I can do to help. (I speak #!/bin/bash)

**Comment by jcisio** [comment]
Thanks for the deep analysis! I'd like to add that writing to disk blocks
everything. There is a standing issue about moving big torrents makes
Transmission unresponsive (no transfer, remote does not work etc.) during
the action.

**Comment by UltimateByte** [comment]
You're welcome @jcisio , trying to help as I can.

Like you, I've also noticed freezes of transmission-remote (and probably the daemon as well) upon completing torrent downloads.
We should maybe open a whole issue about this if nobody did already.

Please, note that I am using the "incomplete" torrent feature (are you using it as well?), so I assumed this sluggishness was due to moving the torrent by copying then removing files. But since my two destinations are on the same partition, it should just run an "mv" of the file which should occur instantly. So if it's not due to a "mv" or "cp / rm", then it might be due to a checksum running on torrent completion.

Whatever the cause, torrent completion shouldn't make the client or torrent server freeze entirely. I would say it doesn't help providing a smooth experience of using Transmission which is very saddening for such a great tool and project.

Project maintainers would likely know more about this. But man, 571 opened issues and almost no tags on it: can they even keep up with the demand? Maybe they need a hand to deal with that Git!

**Comment by aaronjwood** [comment]
FWIW, my experience with 3.0 is very strange https://github.com/transmission/transmission/issues/1314#issuecomment-737651316

There seems to be some starvation issue with the event loop, similar to @zappepappe and BSD's select-based (strange, thought kqueue would have been used) event loop.

**Comment by UltimateByte** [comment]
What I've noticed, is Transmission seems to stop (or at least slow a lot) downloading while it is writing to the disk.
Then when it goes back to downloading, but it takes some time to get back to the max speed. That results in a slow download.
And on top of that, GUI becomes unresponsive as well during writes.

If anything, maybe that is what should be checked.

**Comment by ckerr** [comment]
There seems to be more than one issue being discussed in here.

The 100% CPU issue looks like a duplicate of https://github.com/transmission/transmission/issues/1829.

The "stops downloading while it is writing to the disk" issue is handled in the multithreading issue. Yes, disk IO can be blocking at the moment.

I'm closing this issue since these separate topics are already discussed in different issues. If I've missed anything -- this is a long thread -- please feel free to open new issues to cover them. Thank you!

**Comment by ma5ter** [comment]
The issue is not solved at least in 3.0, has nothing to speed-limiting or webseed, not a duplicate to any other.
Closing issues in bunch without proper investigation is not a good practice/

**Comment by ckerr** [comment]
@ma5ter there are _several_ topics being discussed in this issue so when you say "the issue" I don't know which one you're referring to.

If you have a repeatable, testable issue that is happening in the nightly builds -- I know 3.0 is the newest stable, but the code has undergone a _huge_ rewrite in the interim as part of moving from C to C++, so please don't use 3.0 as  the baseline -- then please open a new ticket describing the issue in detail.
